forte/data/extractor/base_extractor.py

```diff
@@ -0,0 +1,244 @@
+            self.vocab = Vocabulary(method=self.config.vocab_method,
+                                    need_pad=self.config.need_pad,
+                                    use_unk=self.config.vocab_use_unk)
+        else:
+            self.vocab = None
+
+    @staticmethod
+    def default_configs():
+        r"""
+        Returns a dictionary of default hyper-parameters.
+
+        "vocab_method": str
+            What type of vocabulary is used for this extractor.
+            "raw", "indexing", "one-hot" are supported, default is "indexing".
+            Check the behavior of vocabulary under different setting
+            in :class`forte.data.vocabulary.Vocabulary`
+
+        "need_pad": bool
+            Wether the <PAD> element should be added to vocabulary. And
+            wether the feature need to be batched and paded. Default is True.
+
+        "vocab_use_unk": bool
+            Whether the <UNK> element should be added to vocabulary.
+            Default is true.
+        """
+        return {
+            "vocab_method": "indexing",
+            "vocab_use_unk": True,
+            "need_pad": True,
+        }
+
+    @property
+    def state(self) -> Dict:
+        return {
+            "vocab_method": self.config.vocab_method,
+            "need_pad": self.config.need_pad,
+            "vocab_use_unk": self.config.vocab_use_unk,
+            "entry_type": self.config.entry_type,
+            "vocab": self.vocab.state if self.vocab else None,
+        }
+
+    @classmethod
+    def from_state(cls, state: Dict) -> object:
+        config = {
+            "vocab_method": state["vocab_method"],
+            "need_pad": state["need_pad"],
+            "vocab_use_unk": state["vocab_use_unk"],
+            "entry_type": state["entry_type"]
+        }
+        obj = cls(config)
+        if "vocab" in state and state["vocab"] is not None and \
+            isinstance(state["vocab"], dict):
+            obj.vocab = Vocabulary.from_state(state["vocab"])
+        else:
+            obj.vocab = None
+        return obj
+
+    @property
+    def entry_type(self) -> Type[Annotation]:
+        return self.config.entry_type
+
+    @property
+    def vocab_method(self) -> str:
+        return self.config.vocab_method
+
+    def get_pad_value(self) -> Union[None, int, List[int]]:
+        if self.vocab:
+            return self.vocab.get_pad_value()
+        else:
+            return None
+
+    def items(self) -> Iterable[Tuple[Hashable, int]]:
+        assert self.vocab
+        return self.vocab.items()
+
+    def size(self) -> int:
+        assert self.vocab
+        return len(self.vocab)
+
+    def add(self, element: Hashable):
+        assert self.vocab
+        return self.vocab.add_element(element)
+
+    def has_element(self, element: Hashable) -> bool:
+        assert self.vocab
+        return self.vocab.has_element(element)
+
+    def element2repr(self, element: Hashable) -> Union[int, List[int]]:
+        assert self.vocab
```

Comment:
Instead of `assert`, we'd better throw exception with clear error message.

Reply:
I am planning to change it into `assert self.vocab, self.VOCAB_ERROR_MSG`. Is it ok.
Or should I use `if ...: raise Error`?

Reply:
I see. So this do show the error messages. I think throw exceptions is more consistent with the other part of the code.